node/src/store/header_ranges.rs

```diff
@@ -0,0 +1,568 @@
+use std::fmt::Display;
+use std::ops::RangeInclusive;
+use std::vec;
+
+#[cfg(any(test, feature = "test-utils"))]
+use celestia_types::test_utils::ExtendedHeaderGenerator;
+use celestia_types::ExtendedHeader;
+use serde::Serialize;
+use smallvec::{IntoIter, SmallVec};
+
+use crate::store::utils::{ranges_intersection, try_consolidate_ranges, RangeScanResult};
+use crate::store::StoreError;
+
+pub type HeaderRange = RangeInclusive<u64>;
+
+macro_rules! header_ranges {
+    ( $( $r:expr ),* ) => {
+        {
+            $crate::store::header_ranges::HeaderRanges::from_vec(smallvec::smallvec![
+            $(
+                $r,
+            )*
+            ])
+        }
+    };
+}
+
+pub(crate) use header_ranges;
+
+/// Span of header that's been verified internally
+#[derive(Clone)]
+pub struct VerifiedExtendedHeaders(Vec<ExtendedHeader>);
+
+impl IntoIterator for VerifiedExtendedHeaders {
+    type Item = ExtendedHeader;
+    type IntoIter = vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl From<VerifiedExtendedHeaders> for Vec<ExtendedHeader> {
+    fn from(value: VerifiedExtendedHeaders) -> Self {
+        value.into_iter().collect()
+    }
+}
+
+impl AsRef<[ExtendedHeader]> for VerifiedExtendedHeaders {
+    fn as_ref(&self) -> &[ExtendedHeader] {
+        &self.0
+    }
+}
+
+/// 1-length hedaer span is internally verified, this is valid
+impl From<[ExtendedHeader; 1]> for VerifiedExtendedHeaders {
+    fn from(value: [ExtendedHeader; 1]) -> Self {
+        Self(value.into())
+    }
+}
+
+impl From<ExtendedHeader> for VerifiedExtendedHeaders {
+    fn from(value: ExtendedHeader) -> Self {
+        Self(vec![value])
+    }
+}
+
```

Comment:
And remove `.clone()` on the callers

```suggestion

impl<'a> From<&'a ExtendedHeader> for VerifiedExtendedHeaders {
    fn from(value: &ExtendedHeader) -> Self {
        Self(vec![value.to_owned()])
    }
}

```